data_types/src/partition.rs

```diff
@@ -0,0 +1,406 @@
+    /// persist operations when new tags occur in newly inserted data.
+    ///
+    /// Updating inserts new column into the existing order. The order
+    /// of the existing columns relative to each other is NOT changed.
+    ///
+    /// For example, updating `A,B,C` to either `A,D,B,C` or `A,B,C,D`
+    /// is legal. However, updating to `A,C,D,B` is not because the
+    /// relative order of B and C have been reversed.
+    pub sort_key: Vec<String>,
+
+    /// The time at which the newest file of the partition is created
+    pub new_file_at: Option<Timestamp>,
+}
+
+impl Partition {
+    /// Create a new Partition data object from the given attributes. This constructor will take
+    /// care of computing the [`PartitionHashId`].
+    pub fn new(
+        id: PartitionId,
+        table_id: TableId,
+        partition_key: PartitionKey,
+        sort_key: Vec<String>,
+        new_file_at: Option<Timestamp>,
+    ) -> Self {
+        let hash_id = PartitionHashId::new(table_id, &partition_key);
+        Self {
+            id,
+            hash_id: Some(hash_id),
+            table_id,
+            partition_key,
+            sort_key,
+            new_file_at,
+        }
+    }
+
+    /// If this partition has a `PartitionHashId` stored in the catalog, use that. Otherwise, use
+    /// the database-assigned `PartitionId`.
+    pub fn transition_partition_id(&self) -> TransitionPartitionId {
+        self.hash_id
+            .clone()
+            .map(TransitionPartitionId::Deterministic)
+            .unwrap_or_else(|| TransitionPartitionId::Deprecated(self.id))
+    }
+
+    /// The unique hash derived from the table ID and partition key, if it exists in the catalog.
+    pub fn hash_id(&self) -> Option<&PartitionHashId> {
+        self.hash_id.as_ref()
+    }
+
+    /// The sort key for the partition, if present, structured as a `SortKey`
+    pub fn sort_key(&self) -> Option<SortKey> {
+        if self.sort_key.is_empty() {
+            return None;
+        }
+
+        Some(SortKey::from_columns(self.sort_key.iter().map(|s| &**s)))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use proptest::{prelude::*, proptest};
+
+    #[test]
+    fn display_partition_hash_id_in_hex() {
+        let partition_hash_id =
+            PartitionHashId::new(TableId::new(5), &PartitionKey::from("2023-06-08"));
+
+        assert_eq!(
+            "ebd1041daa7c644c99967b817ae607bdcb754c663f2c415f270d6df720280f7a",
+            partition_hash_id.to_string()
+        );
+    }
+
+    proptest! {
+        #[test]
+        fn partition_hash_id_representations(
+            table_id in 0..i64::MAX,
+            partition_key in any::<String>(),
+        ) {
+            prop_assume!(!partition_key.is_empty());
+
+            let table_id = TableId::new(table_id);
+            let partition_key = PartitionKey::from(partition_key);
+
+            let partition_hash_id = PartitionHashId::new(table_id, &partition_key);
```

Comment:
How about throwing this in:

```suggestion
            let partition_hash_id = PartitionHashId::new(table_id, &partition_key);
            
            // ID generation MUST be deterministic.
            {
                let partition_hash_id2 = PartitionHashId::new(table_id, &partition_key);
                assert_eq!(partition_hash_id, partition_hash_id2);
            }
```

If you're feeling spicy, you _could_ also do:

```rust
// ID generation MUST be collision resistant; different inputs -> different IDs
{
    let other_table_id = TableId::new(table_id.get().wrapping_add(1));
    let partition_hash_id3 = PartitionHashId::new(other_table_id, &partition_key);
    assert_ne!(partition_hash_id, partition_hash_id3);
}
```

Which whilst being a useful property assert of its own, also ensures we're not just returning `bananas` as every ID.